Prometheus servicemonitor CRD example failing

Used provided example for prometheus. It's crashlooping like this:

```
I1115 02:10:36.814960       1 logr.go:252] plugin: Waiting for vcluster to become leader...
I1115 02:10:36.815568       1 logr.go:252] plugin: Starting syncers...
I1115 02:10:36.916286       1 logr.go:252] plugin: Start syncer ServiceMonitor-from-virtual-syncer
I1115 02:10:36.916685       1 logr.go:252] plugin: Successfully started plugin.
I1115 02:10:36.916892       1 logr.go:252] controller ServiceMonitor-from-virtual-syncer controllerGroup monitoring.coreos.com controllerKind ServiceMonitor: Starting EventSource source kind source: *unstructured.Unstructured
I1115 02:10:36.916946       1 logr.go:252] controller ServiceMonitor-from-virtual-syncer controllerGroup monitoring.coreos.com controllerKind ServiceMonitor: Starting EventSource source &{{%!s(*unstructured.Unstructured=&{map[apiVersion:monitoring.coreos.com/v1 kind:ServiceMonitor]}) %!s(*cache.informerCache=&{0xc00007eec0}) %!s(chan error=<nil>) %!s(func()=<nil>)}}
I1115 02:10:36.916960       1 logr.go:252] controller ServiceMonitor-from-virtual-syncer controllerGroup monitoring.coreos.com controllerKind ServiceMonitor: Starting Controller
I1115 02:10:37.017707       1 logr.go:252] controller ServiceMonitor-from-virtual-syncer controllerGroup monitoring.coreos.com controllerKind ServiceMonitor: Starting workers worker count 1
I1115 02:10:37.017827       1 logr.go:252] ServiceMonitor-from-virtual-syncer: insights-discovery-proxy: Create physical ServiceMonitor i56yke0idw/insights-discovery-proxy, since it is missing, but virtual object exists
I1115 02:10:37.018181       1 logr.go:252] controller ServiceMonitor-from-virtual-syncer controllerGroup monitoring.coreos.com controllerKind ServiceMonitor: serviceMonitor klog.ObjectRef{Name:"insights-discovery-proxy", Namespace:"i56yke0idw"} namespace i56yke0idw name insights-discovery-proxy: reconcileID "98226c3c-cda0-4ce1-8313-f82110b1b2a3": : Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x123a2e5]

goroutine 465 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x1826e80, 0x2992910})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.replaceChildAtIndex(...)
	/go/vcluster/pkg/patches/operation.go:70
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.ReplaceNode(0x175db40?, 0xc000ac1540, 0xc000ac2f00)
	/go/vcluster/pkg/patches/operation.go:11 +0x65
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.ValidateAndTranslateName(0x1?, 0xc000ac1540, 0xc00069e320, {0x1cefb18, 0xc0006be620}, {0x0, 0x0})
	/go/vcluster/pkg/patches/patch_types.go:261 +0x145
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.ProcessRewrite(0x123a580?, 0xc00069e320, {0x1cefb18, 0xc0006be620})
	/go/vcluster/pkg/patches/patch_types.go:210 +0x10c
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.RewriteName(0xa80?, 0xc00069e320, {0x1cefb18, 0xc0006be620})
	/go/vcluster/pkg/patches/patch_types.go:181 +0x145
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.applyPatch(0x1a571e0?, 0xc000466080?, 0x7fbe6a3adc20?, {0x1cefb18?, 0xc0006be620?})
	/go/vcluster/pkg/patches/patch.go:78 +0x125
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/patches.ApplyPatches({0x1cf9750?, 0xc000466080}, {0x0?, 0x0}, {0xc00012cf40, 0x7, 0xc0006be620?}, {0x0, 0x0, 0x0}, ...)
	/go/vcluster/pkg/patches/patch.go:41 +0x175
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/syncer.(*patcher).ApplyPatches(0xc00012d280, {0x1cedbc8, 0xc000679830}, {0x1cf9750?, 0xc000466048?}, {0x0, 0x0}, {0xc00012cf40, 0x7, 0x7}, ...)
	/go/vcluster/pkg/syncer/generic.go:44 +0x105
github.com/loft-sh/vcluster-generic-crd-plugin/pkg/syncer.(*fromVirtualController).SyncDown(0xc0001aa600, 0xc000653110, {0x1cf9750?, 0xc000466048})
	/go/vcluster/pkg/syncer/from_virtual_syncer.go:84 +0x337
github.com/loft-sh/vcluster-sdk/syncer.(*syncerController).Reconcile(0xc000652fc0, {0x1cedbc8?, 0xc000679830}, {{{0xc0008ebbc0?, 0x10?}, {0xc00063bb18?, 0x40d7e7?}}})
	/go/vcluster/vendor/github.com/loft-sh/vcluster-sdk/syncer/syncer.go:96 +0x5c3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1cedb20?, {0x1cedbc8?, 0xc000679830?}, {{{0xc0008ebbc0?, 0x1990760?}, {0xc00063bb18?, 0x4041f4?}}})
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000c46e0, {0x1cedb20, 0xc000305ec0}, {0x189dfc0?, 0xc00007f360?})
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000c46e0, {0x1cedb20, 0xc000305ec0})
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/vcluster/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:230 +0x325
```

--- a/devspace.yaml
+++ b/devspace.yaml
@@ -35,6 +35,9 @@ deployments:
                   - apiGroups: ["cert-manager.io"]
                     resources: ["issuers", "certificates", "certificaterequests"]
                     verbs: ["create", "delete", "patch", "update", "get", "list", "watch"]
+                  - apiGroups: ["monitoring.coreos.com"]
+                    resources: ["servicemonitors"]
+                    verbs: ["create", "delete", "patch", "update", "get", "list", "watch"]
             env:
               # Config for testing
               # run `make install-cert-manager` within the host kube context before using
@@ -59,6 +62,10 @@ deployments:
 dev:
   vcluster:
     imageSelector: ${PLUGIN_IMAGE}
+    resources:
+      limits:
+        cpu: "0" # to prevent CPU throttling during compilation
+        memory: "0" # to prevent OOM kills which can happen during compilation
     terminal:
       command: "./devspace_start.sh"
     ports:

--- a/examples/prometheus-operator/Readme.md
+++ b/examples/prometheus-operator/Readme.md
@@ -1,7 +1,7 @@
 # How to use vcluster-generic-crd-sync-plugin to sync Prometheus Operator resources
 **Note**: Currently this plugin is still underdevelopment and is provided without any commercial support. You are encouraged to report problems via Issues in this repo. For the help with troubleshooting you can post to the "vcluster" channel in our Slack - https://slack.loft.sh/  
-**Note**: The configuration provided here is an example. Please review and update it to match your environment and use case.
-
+**Note**: The configuration provided here is an example. Please review and update it to match your environment and use case.  
+**Warning**: namespace selectors are currently not supported, so with the provided configuration the `.spec.namespaceSelector` field is ignored and Services from all vcluster namespaces are selected. You can upvote [issue #15](https://github.com/loft-sh/vcluster-generic-crd-sync-plugin/issues/15) if the namespace selectors are critical for your use cases.
 
 Prerequisite for installation are the Prometheus CRDs installed in the host cluster where vcluster will be installed.
 

--- a/examples/prometheus-operator/plugin.yaml
+++ b/examples/prometheus-operator/plugin.yaml
@@ -37,8 +37,6 @@ plugin:
                     path: .spec.targetLabels[*]
                   - op: rewriteLabelKey
                     path: .spec.podTargetLabels[*]
-                  - op: rewriteName
-                    path: .spec.endpoints[*]
                   - op: rewriteLabelExpressionsSelector
                     path: .spec.selector
                     # TODO: Dev: add transformed namespace selectors to the label selectors (.spec.selector)
@@ -48,3 +46,57 @@ plugin:
                     value:
                       any: false
                       matchNames: []
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.ca.secret
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.ca.configMap
+                    sync:
+                      configmap: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.cert.secret
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.cert.configMap
+                    sync:
+                      configmap: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.bearerTokenSecret.name
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.authorization.credentials.name
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.basicAuth.username.name
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.basicAuth.password.name
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.oauth2.clientId.secret
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.oauth2.clientId.configMap
+                    sync:
+                      configmap: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.oauth2.clientSecret.name
+                    sync:
+                      secret: true
+                  - op: rewriteName
+                    path: .spec.endpoints[*].tlsConfig.oauth2.tokenUrl
+                    # regex below should translate service urls only
+                    regex: >
+                      ^http(s)?://$NAME((\.$NAMESPACE)?(\.svc(\.cluster\.local)?){1})?(/|$)
+                  - op: rewriteName
+                    path: .spec.endpoints[*].proxyUrl
+                    # regex below should translate service urls only
+                    regex: >
+                      ^http(s)?://$NAME((\.$NAMESPACE)?(\.svc(\.cluster\.local)?){1})?(/|$)

--- a/pkg/patches/operation.go
+++ b/pkg/patches/operation.go
@@ -8,7 +8,9 @@ import (
 
 func ReplaceNode(doc *yaml.Node, match *yaml.Node, value *yaml.Node) {
 	parent := Find(doc, ContainsChild(match))
-	parent.Content = replaceChildAtIndex(parent, match, value)
+	if parent != nil {
+		parent.Content = replaceChildAtIndex(parent, match, value)
+	}
 }
 
 func Find(doc *yaml.Node, predicate func(*yaml.Node) bool) *yaml.Node {

--- a/pkg/patches/patch_test.go
+++ b/pkg/patches/patch_test.go
@@ -397,6 +397,22 @@ test2: {}`,
         - vcluster.loft.sh/label-suffix-x-cb4e76426f
         - vcluster.loft.sh/label-suffix-x-bae4a2c2e5`,
 		},
+		{
+			name: "rewrite name should not panic when match is not scalar",
+			patch: &config.Patch{
+				Operation: config.PatchTypeRewriteName,
+				Path:      "test.endpoints[*]",
+			},
+			nameResolver: &fakeVirtualToHostNameResolver{},
+			obj1: `test:
+    endpoints: 
+      - name: abc
+      - name: def`,
+			expected: `test:
+    endpoints:
+        - name: abc
+        - name: def`,
+		},
 	}
 
 	for _, testCase := range testCases {

--- a/pkg/patches/patch_types.go
+++ b/pkg/patches/patch_types.go
@@ -207,6 +207,9 @@ func ProcessRewrite(obj *yaml.Node, patch *config.Patch, resolver NameResolver) 
 	}
 
 	for _, nameMatch := range nameMatches {
+		if nameMatch.Kind != yaml.ScalarNode {
+			continue
+		}
 		err = ValidateAndTranslateName(obj, nameMatch, patch, resolver, namespace)
 
 		if err != nil {
@@ -222,6 +225,9 @@ func ProcessRewrite(obj *yaml.Node, patch *config.Patch, resolver NameResolver) 
 		}
 
 		for _, namespaceMatch := range namespaceMatches {
+			if namespaceMatch.Kind != yaml.ScalarNode {
+				continue
+			}
 			err = ValidateAndTranslateNamespace(obj, namespaceMatch, patch, resolver)
 
 			if err != nil {

